### PR TITLE
Fit brms

### PR DESCRIPTION
**Pull request by wlangera**
-  Fitting simpel models with brms
-  Comparing models
-  Post processing results
-  Comparison with frequentist framework

### CODE REVIEW

File: source/workshop_1_mcmc_en_brms.Rmd [hunk 1]
```diff
@@ -77,201 +68,530 @@ ants_df <- ants %>%
...
+  as.data.frame() %>%
+  select(elpd_diff, se_diff) %>%
+  mutate(ll_diff = elpd_diff  + qnorm(0.025) * se_diff,
+         ul_diff = elpd_diff  + qnorm(0.975) * se_diff)
+```
+
+## K-fold cross-validation
 
-https://mc-stan.org/loo/articles/online-only/faq.html
+Bij K-fold cross-validation wordt de data in $K$ groepen opgesplitst. Wij zullen hier $K = 10$ groepen (= folds) gebruiken. In plaats van dus telkens één enkele observatie weg te laten zoals bij leave-one-out CV gaan we hier $1/10$e van de data weglaten. Via de argumenten `folds = "stratified"` en `group = "habitat"` zorgen we ervoor dat voor elke groep de relatieve frequenties van habitat bewaard blijven. Deze techniek zal dus minder precies zijn dan de vorige, maar zal sneller zijn om te berekenen indien je met heel veel data werkt.
```
Comment (wlangera) [inline source/workshop_1_mcmc_en_brms.Rmd:501]:
```suggestion
Bij K-fold cross-validation worden de data in $K$ groepen opgesplitst. Wij zullen hier $K = 10$ groepen (= folds) gebruiken. In plaats van dus telkens één enkele observatie weg te laten zoals bij leave-one-out CV gaan we hier $1/10$e van de data weglaten. Via de argumenten `folds = "stratified"` en `group = "habitat"` zorgen we ervoor dat voor elke groep de relatieve frequenties van habitat bewaard blijven. Deze techniek zal dus minder precies zijn dan de vorige, maar zal sneller zijn om te berekenen indien je met heel veel data werkt.
```

Comment (wlangera) [inline source/workshop_1_mcmc_en_brms.Rmd]:
```suggestion
Bij K-fold cross-validation worden de data in $K$ groepen opgesplitst. Wij zullen hier $K = 10$ groepen (= folds) gebruiken. In plaats van dus telkens één enkele observatie weg te laten zoals bij leave-one-out CV gaan we hier $1/10$e van de data weglaten. Via de argumenten `folds = "stratified"` en `group = "habitat"` zorgen we ervoor dat voor elke groep de relatieve frequenties van habitat bewaard blijven. Deze techniek zal dus minder precies zijn dan de vorige, maar zal sneller zijn om te berekenen indien je met heel veel data werkt.
```

File: source/workshop_1_mcmc_en_brms.Rmd [hunk 2]
```diff
@@ -77,201 +68,530 @@ ants_df <- ants %>%
...
 
+Als oefening zullen we een model maken om het aantal soorten te vergelijken tussen beide habitats. Uit de data exploratie zagen we al dat het aantal hoger lijkt te liggen in bossen en dat sites met een hoger aantal in moerassen vaak ook een hoger aantal in bossen hebben.
+
+## Specificatie van een lineaire regressie
 
-## Een eerste simpel model
 ### Model specificatie
 
-Eerst en vooral specificeren we de MCMC parameters ...
+We willen een model waarmee we het verschil kunnen onderzoeken in aantal soorten mieren tussen moeras- en boshabitats. Beschouw de responsvariabele $Y$ het aantal mieren en $X_{habitat}$ een dummy variabele die gelijk is aan 0 voor moerassen en 1 voor bossen. We veronderstellen dat $Y$ een Normaal verdeling volgt (equivalent aan een lineaire regressie met categorische variabele (= ANOVA))
```
Comment (RCinbo) [inline source/workshop_1_mcmc_en_brms.Rmd]:
```suggestion
We willen een model waarmee we het verschil kunnen onderzoeken in aantal soorten mieren tussen moeras- en boshabitats. Beschouw de responsvariabele $Y$, het aantal mieren, en $X_{habitat}$, een dummy variabele die gelijk is aan 0 voor moerassen en 1 voor bossen. We veronderstellen dat $Y$ een Normaal verdeling volgt (equivalent aan een lineaire regressie met categorische variabele (= ANOVA))
```

File: source/workshop_1_mcmc_en_brms.Rmd [hunk 3]
```diff
@@ -77,201 +68,530 @@ ants_df <- ants %>%
...
 nparallel <- nchains   # aantal cores voor parallel computing
-thinning <- 1          # verdunningsfactor (hier 1 =geen verdunning)
+thinning <- 1          # verdunningsfactor (hier 1 = geen verdunning)
 ```
 
-$$
-Y \sim Pois(\beta_0 + \beta_1X_{habitat})\\
-\Rightarrow \ln(E(Y)) = \beta_0 + \beta_1X_{habitat}
-$$
+Het model wordt gefit a.d.h.v. de `brm()` functie. De syntax is zeer gelijkaardig als functies die in frequentist statistics worden gebruikt zoals `lm()` en `glm()`. De `brm()` functie bevat heel wat argumenten (zie `?brm`). Handige argumenten die we hier niet gebruiken zijn bijvoorbeeld:
```
Comment (RCinbo) [inline source/workshop_1_mcmc_en_brms.Rmd]:
```suggestion
Het model wordt gefit a.d.h.v. de `brm()` functie. De syntax is zeer gelijkaardig aan functies die in frequentist statistics worden gebruikt zoals `lm()` en `glm()`. De `brm()` functie bevat heel wat argumenten (zie `?brm`). Handige argumenten die we hier niet gebruiken zijn bijvoorbeeld:
```

File: source/workshop_1_mcmc_en_brms.Rmd [hunk 4]
```diff
@@ -77,201 +68,530 @@ ants_df <- ants %>%
...
+
+-   `inits` om de beginwaarden van MCMC algoritme in te stellen. Indien deze niet gespecificeerd zijn, gebruikt de functie default waarden.
+-   `prior` om prior distributies te voorzien. Indien deze niet gespecificeerd zijn, gebruikt de functie default (niet informatieve) priors. `sample_prior` kan gebruikt worden om van de prior te samplen zodat je kan visualiseren of de gebruikte prior voldoet aan je verwachtingen.
```
Comment (RCinbo) [inline source/workshop_1_mcmc_en_brms.Rmd]:
```suggestion
-   `prior` om prior distributies te voorzien. Indien deze niet gespecificeerd zijn, gebruikt de functie default (niet informatieve) priors. Het argument `sample_prior` kan gebruikt worden om van de prior te samplen zodat je kan visualiseren of de gebruikte prior voldoet aan je verwachtingen.
```

File: source/workshop_1_mcmc_en_brms.Rmd [hunk 5]
```diff
@@ -77,201 +68,530 @@ ants_df <- ants %>%
...
+  data = ants_df,              # ingeven data
+  chains = nchains,            # MCMC parameters
+  warmup = burnin, 
+  iter = niter,
+  cores = nparallel,
+  thin = thinning,
+  seed = 123)                  # seed voor reproduceerbare uitkomst
 ```
 
+Voor we de resultaten bekijken, controleren we eerst of het model goed geconvergeerd is.
```
Comment (RCinbo) [inline source/workshop_1_mcmc_en_brms.Rmd]:
```suggestion
Voor we de resultaten bekijken, controleren we eerst of het model goed convergeert.
```

File: source/workshop_1_mcmc_en_brms.Rmd [hunk 6]
```diff
@@ -77,201 +68,530 @@ ants_df <- ants %>%
...
 
-Wat is een trace plot?
+```{r}
+# Welke parameters gaan we bekijken?
+parameters <- c("b_Intercept", "b_habitatForest", "sigma")
+```
+
+**trace plot**
+
+Trace plots geven aan hoe het samplen van de posterior distribution verloopt over de tijd. Het idee is dat elke chain convergeert naar dezelfde dezelfde distributie voor elke parameter (we spreken van mixing). Als deze plot een harige rups ('fuzzy caterpillar') voorstelt, wil dit zeggen dat convergentie goed is.
```
Comment (RCinbo) [inline source/workshop_1_mcmc_en_brms.Rmd]:
```suggestion
Trace plots geven aan hoe het samplen van de posterior distribution verloopt over de tijd. Het idee is dat elke chain convergeert naar dezelfde distributie voor elke parameter (we spreken van mixing). Als deze plot eruit ziet als een harige rups ('fuzzy caterpillar'), wil dit zeggen dat de convergentie goed is.
```

File: source/workshop_1_mcmc_en_brms.Rmd [hunk 7]
```diff
@@ -77,201 +68,530 @@ ants_df <- ants %>%
...
 ```{r}
-mcmc_trace(fit1, pars = c("b_Intercept", "b_habitatForest"))
+# Visualisatie via Bayesplot package
+mcmc_trace(fit_normal1, pars = parameters)
 ```
 
-Wat is posterior density?
+**running mean/quantile/... plot**
+
+Gelijkaardig als een trace plot kunnen we ook kijken hoe bepaalde statistieken zoals het gemiddelde of de quantielen zich voordoen over de tijd (= met toename aantal iteraties). Je wilt natuurlijk dat deze statistieken ook constant zijn na een aantal iteraties. Dan weet je dat je genoeg iteraties gebruikt hebt.
```
Comment (RCinbo) [inline source/workshop_1_mcmc_en_brms.Rmd]:
```suggestion
Naast een trace plot, kunnen we ook kijken hoe bepaalde statistieken zoals het gemiddelde of de kwantielen variëren over de tijd (= met toename aantal iteraties). Je wilt natuurlijk dat deze statistieken stabiliseren na een aantal iteraties want dan weet je dat je genoeg iteraties gebruikt hebt.
```

File: source/workshop_1_mcmc_en_brms.Rmd [hunk 8]
```diff
@@ -77,201 +68,530 @@ ants_df <- ants %>%
...
 
 ```{r}
-mcmc_dens_overlay(fit1, pars = c("b_Intercept", "b_habitatForest"))
+# code om cumulatieve quantielen te berekenen
```
Comment (RCinbo) [inline source/workshop_1_mcmc_en_brms.Rmd]:
```suggestion
# code om cumulatieve kwantielen te berekenen
```

File: source/workshop_1_mcmc_en_brms.Rmd [hunk 9]
```diff
@@ -77,201 +68,530 @@ ants_df <- ants %>%
...
+  # splits naam kolom op bij laatste underscore
+  extract(name, into = c("parameter", "statistiek"), "(.*)_([^_]+)$") %>%
+  ggplot(aes(y = value, x = .iteration, colour = factor(.chain))) +
+    geom_line(aes(linetype = statistiek), linewidth = 0.8) +
+    facet_wrap(~parameter, nrow = 3, scales = "free")
+```
+
+**posterior density plot**
+
+De vorm van de posterior distributions kan ook informatief zijn of de chains geconvergeerd zijn naar dezelfde distributies. Bovendien is het bijvoorbeeld mogelijk dat er meerdere maxima aanwezig zijn. Dan zou er mogelijks ook iets mis kunnen zijn met de specificatie van ons model.
```
Comment (RCinbo) [inline source/workshop_1_mcmc_en_brms.Rmd]:
```suggestion
De vorm van de posterior distributions kan ook informatief zijn om na te kijken of de chains geconvergeerd zijn naar dezelfde distributies. Als we een bimodale verdeling zien (een kamelenrug met twee pieken), dan zou er mogelijks ook iets mis kunnen zijn met de specificatie van ons model.
```

File: source/workshop_1_mcmc_en_brms.Rmd [hunk 10]
```diff
@@ -77,201 +68,530 @@ ants_df <- ants %>%
...
+ratios_fit_normal1 <- neff_ratio(fit_normal1)[parameters]
+print(ratios_fit_normal1)
+```
+
+```{r}
+# Plot verhoudingen via Bayesplot package
+mcmc_neff(ratios_fit_normal1) + yaxis_text(hjust = 1)
+```
+
+Andere packages die van pas kunnen komen bij het controleren van MCMC convergentie zijn [mcmcplots](https://cran.r-project.org/web/packages/mcmcplots/index.html) en [coda](https://cran.r-project.org/web/packages/coda/index.html). Voor deze packages moet je de MCMC samples wel eerst omzetten naar een mcmc object.
```
Comment (RCinbo) [inline source/workshop_1_mcmc_en_brms.Rmd]:
```suggestion
Andere packages die van pas kunnen komen bij het controleren van MCMC convergentie zijn [mcmcplots](https://cran.r-project.org/web/packages/mcmcplots/index.html) en [coda](https://cran.r-project.org/web/packages/coda/index.html). Voor deze packages moet je de MCMC samples uit je **brms** model wel eerst omzetten naar een mcmc object.
```

File: source/workshop_1_mcmc_en_brms.Rmd [hunk 11]
```diff
@@ -77,201 +68,530 @@ ants_df <- ants %>%
...
+install.packages(mcmcplots)
+mcmcplots::mcmcplot(as.mcmc(fit_normal1, pars = parameters))
+```
+
+We kunnen over het algemeen besluiten dat MCMC convergentie goed is voor alle parameters. Indien dit niet het geval zou zijn, kan het zijn dat de MCMC parameters anders moeten gekozen worden, bijvoorbeeld door het aantal iteraties te verhogen. Nu we weten dat de parameters correct geschat zijn, kunnen we controleren of het model goed bij de data past.
+
 ### Model fit
 
-https://mc-stan.org/bayesplot/articles/graphical-ppcs.html
+De [posterior predictive check](https://mc-stan.org/bayesplot/articles/graphical-ppcs.html) (PPC) is een goede manier om te controleren of het model goed past bij de data. Het idee achter de PPC is dat, als een model goed past, de voorspelde waardes o.b.v. het model veel lijken op de data die we hebben gebruikt om het model te fitten. Om de voorspellingen te genereren die worden gebruikt voor PPCs, simuleren we vanuit de posterior predictive distribution (de posterior distribution van de responsvariabele). Visualisatie kan met de **bayesplot** package. De dikke lijn geeft de data aan en de dunne lijntjes verschillende simulaties o.b.v. het model.
```
Comment (RCinbo) [inline source/workshop_1_mcmc_en_brms.Rmd]:
```suggestion
De [posterior predictive check](https://mc-stan.org/bayesplot/articles/graphical-ppcs.html) (PPC) is een goede manier om te controleren of het model goed past bij de data. Het idee achter de PPC is dat, als een model goed past, de voorspelde waardes o.b.v. het model veel lijken op de data die we hebben gebruikt om het model te fitten. Om de voorspellingen te genereren die worden gebruikt voor PPCs, simuleren we vanuit de posterior predictive distribution (de posterior distribution van de responsvariabele). Visualisatie kan met de **bayesplot** package. De dikke lijn toont de data en de dunne lijntjes tonen verschillende simulaties o.b.v. het model.
```

File: source/workshop_1_mcmc_en_brms.Rmd [hunk 12]
```diff
@@ -77,201 +68,530 @@ ants_df <- ants %>%
...
+
+```{r}
+# Visualiseer model fit via Bayesplot package
+pp_check(fit_normal1, type = "dens_overlay_grouped", ndraws = 100, 
+         group = "habitat")
+```
+
+We zien dat de overeenkomst tussen de geobserveerde data en de voorspellingen o.b.v. ons model niet helemaal overeenkomen. Mogelijks was onze keuze voor de Normaal verdeling niet goed. Aantallen zijn discrete, positieve waarden, terwijl de Normaal verdeling continue en zowel positieve als negatieve waarden kan voorspellen. De Poisson verdeling is een discrete verdeling die steeds positief is. Ze wordt vaak gebruikt voor het modelleren van geregistreerde aantallen gedurende een gegeven tijdsinterval, afstand, oppervlakte, volume ...
```
Comment (RCinbo) [inline source/workshop_1_mcmc_en_brms.Rmd]:
```suggestion
We zien dat de geobserveerde data en de voorspellingen o.b.v. ons model niet helemaal overeenkomen. Mogelijks was onze keuze voor de Normaal verdeling niet goed. Aantallen zijn discrete, positieve waarden, terwijl de Normaal verdeling continue en zowel positieve als negatieve waarden kan voorspellen. De Poisson verdeling is een discrete verdeling die steeds positief is. Ze wordt vaak gebruikt voor het modelleren van geregistreerde aantallen gedurende een gegeven tijdsinterval, afstand, oppervlakte, volume ...
```

File: source/workshop_1_mcmc_en_brms.Rmd [hunk 13]
```diff
@@ -77,201 +68,530 @@ ants_df <- ants %>%
...
-## Model selectie
+Hoe kunnen we deze modellen nu objectief gaan vergelijken?
+
+# Vergelijken van modellen
+
+Op basis van de PPCs kunnen we reeds zien welk model het best past bij de data. Verder zijn er nog enkele functies die **brms** voorziet om verschillende modellen te vergelijken. Met de functie `add_criterion()` kan je model fit criteria toevoegen aan model objecten. Typ `?add_criterion()` om te zien welke beschikbaar zijn. Zie ook <https://mc-stan.org/loo/articles/online-only/faq.html>
+
+## Leave-one-out cross-validation
+
+Cross-validation CV is een familie van technieken die probeert in te schatten hoe goed een model onbekende data zou voorspellen via predicties van het model gefit op de bekende data. Hiervoor moet je niet per se nieuwe data gaan inzamelen. Je kan jouw eigen data opsplitsen in een test en training dataset. Je fit het model op de training dataset en je gebruikt dat model dan om te schatten hoe goed het de data in de test dataset kan voorspellen. Bij leave-one-out CV ga je telkens één observatie weglaten en het model opnieuw fitten o.b.v. alle andere data.
```
Comment (RCinbo) [inline source/workshop_1_mcmc_en_brms.Rmd]:
```suggestion
Cross-validation (CV) is een familie van technieken die probeert in te schatten hoe goed een model onbekende data zou voorspellen via predicties van het model gefit op de bekende data. Hiervoor moet je niet per se nieuwe data gaan inzamelen. Je kan jouw eigen data opsplitsen in een test en training dataset. Je fit het model op de training dataset en je gebruikt dat model dan om te schatten hoe goed het de data in de test dataset kan voorspellen. Bij leave-one-out CV (LOOCV) ga je telkens één observatie weglaten en het model opnieuw fitten o.b.v. alle andere observaties.
```

File: source/workshop_1_mcmc_en_brms.Rmd [hunk 14]
```diff
@@ -77,201 +68,530 @@ ants_df <- ants %>%
...
+                              criterion = c("loo"))
+fit_poisson2 <- add_criterion(fit_poisson2,
+                              criterion = c("loo"))
+```
+
+Om het verschil tussen de voorspelling van het model o.b.v. de training dataset te vergelijken met de weggelaten data, moeten we een zgn. utitily of loss function definiëren. Wij gebruiken hier de 'expected log pointwise predictive density' (ELPD). Voor deze workshop is het vooral van belang te beseffen dat dit een maat is voor hoe goed jouw Bayesiaanse model nieuwe datapunten voorspelt. Het wordt berekend door de log-likelihood van elk datapunt te nemen uit de posterior predictive distribution van je model en deze vervolgens uit te middelen. Een hogere elpd duidt op een betere modelfit en voorspellende nauwkeurigheid. Voor deze workshop is het vooral van belang te beseffen dat dit een maat is voor hoe goed het model onbekende data kan voorspellen (`elpd_loo` en de standaard error `se_elpd_loo`). `p_loo` is een maat voor de complexiteit van het model. $\text{looic} = -2*\text{elpd_loo}$ Met de functie `loo_compare()` kan je meerdere modellen vergelijken en wordt ook het verschil in ELPD berekend.
```
Comment (RCinbo) [inline source/workshop_1_mcmc_en_brms.Rmd]:
```suggestion
Om het verschil tussen de voorspelling van het model o.b.v. de training dataset te vergelijken met de weggelaten data, moeten we een zgn. utitily of loss function definiëren. Wij gebruiken hier de 'expected log pointwise predictive density' (ELPD). Voor deze workshop is het vooral van belang te beseffen dat dit een maat is voor hoe goed jouw Bayesiaanse model nieuwe datapunten voorspelt. Het wordt berekend door de log-likelihood van elk datapunt te nemen uit de posterior predictive distribution van je model en deze vervolgens uit te middelen. Een hogere ELPD duidt op een betere model fit en voorspellende nauwkeurigheid. Voor deze workshop is het vooral van belang te beseffen dat dit een maat is voor hoe goed het model onbekende data kan voorspellen (`elpd_loo` en de standaard error `se_elpd_loo`). `p_loo` is een maat voor de complexiteit van het model. $\text{looic} = -2*\text{elpd_loo}$ Met de functie `loo_compare()` kan je meerdere modellen vergelijken en wordt ook het verschil in ELPD berekend.
```

File: source/workshop_1_mcmc_en_brms.Rmd [hunk 15]
```diff
@@ -77,201 +68,530 @@ ants_df <- ants %>%
...
+
+```{r}
+fit_poisson2
+```
+
+Een handige package voor visualisatie van de resultaten van ons finale model is de [tidybayes](https://mjskay.github.io/tidybayes/articles/tidy-brms.html) package. Via deze package kan je werken met de posterior distributions zoals je met elke dataset zou werken via het **tidyverse** package.
+
+Met de functie `gather_draws()` kan je een bepaald aantal samples van de posterior distributies van bepaalde parameters gaan nemen en omzetten in een tabel in lang formaat. Je wilt meestal niet alle posterior samples gaan selecteren omdat dit er vaak onnodig veel zijn. Door een `seed` te specificeren zorg je ervoor dat dit telkens dezelfde samples zijn als je het script opnieuw runt. Via de klassieke **dplyr** functies kan je dan bepaalde samenvattende statistieken berekenen.
```
Comment (RCinbo) [inline source/workshop_1_mcmc_en_brms.Rmd]:
```suggestion
Met de functie `gather_draws()` kan je een bepaald aantal samples van de posterior distributies van bepaalde parameters nemen en omzetten in een tabel in lang formaat. Je wilt meestal niet alle posterior samples selecteren omdat dit er vaak onnodig veel zijn. Door een `seed` te specificeren zorg je ervoor dat dit telkens dezelfde samples zijn als je het script opnieuw runt. Via de klassieke **dplyr** functies kan je dan bepaalde samenvattende statistieken berekenen.
```

File: source/workshop_1_mcmc_en_brms.Rmd [hunk 16]
```diff
@@ -281,59 +601,60 @@ fit2 %>%
             max = max(.value))
 ```
 
-Handige functies zijn ook `median_qi()`, `mean_qi()` ...  na `gather_draws()` in plaats van `group_by()` en `summarise()`.
+Handige functies van de **tidybayes** package zijn ook `median_qi()`, `mean_qi()` ... na `gather_draws()` die je kan gebruiken in plaats van `group_by()` en `summarise()`.
 
-We zouden graag het geschatte aantal soorten visualiseren per habitattype.
-Het gemiddeld aantal soorten in moerassen volgens ons model is $\exp(\beta_0)$ en in bossen $\exp(\beta_0+\beta_1)$.
+We zouden nu graag het geschatte aantal soorten visualiseren per habitattype met bijbehorende onzekerheid. Met de functie `spread_draws()` kan je een bepaald aantal samples van de posterior distributie gaan nemen en omzetten in een tabel in wijd formaat. Het gemiddeld aantal soorten in moerassen volgens ons model is $\exp(\beta_0)$ en in bossen $\exp(\beta_0+\beta_1)$. We tonen de posterior distributions met de posterior mediaan en 60 en 90 % credible intervals.
```
Comment (RCinbo) [inline source/workshop_1_mcmc_en_brms.Rmd]:
```suggestion
We zouden nu graag het geschatte aantal soorten visualiseren per habitattype met bijbehorende onzekerheid. Met de functie `spread_draws()` kan je een bepaald aantal samples van de posterior distributie nemen en omzetten in een tabel in wijd formaat. Het gemiddeld aantal soorten in moerassen volgens ons model is $\exp(\beta_0)$ en in bossen $\exp(\beta_0+\beta_1)$. We tonen de posterior distributions met de posterior mediaan en 60 en 90 % credible intervals.
```

**Review by RCinbo — APPROVED** [review]
Looks great, thanks Ward!

**Pull request merged**